vertx-web/src/main/java/io/vertx/ext/web/handler/impl/OAuth2AuthHandlerImpl.java

```diff
@@ -74,54 +92,88 @@ public void handle(RoutingContext ctx) {
-
-              ((OAuth2Auth) authProvider).decodeToken(credentials, decodeToken -> {
-                if (decodeToken.failed()) {
-                  ctx.response().putHeader("WWW-Authenticate", "Bearer error=\"invalid_token\" error_message=\"" + decodeToken.cause().getMessage() + "\"");
-                  ctx.fail(401);
-                  return;
-                }
-
-                ctx.setUser(decodeToken.result());
-                Session session = ctx.session();
-                if (session != null) {
-                  // the user has upgraded from unauthenticated to authenticated
-                  // session should be upgraded as recommended by owasp
-                  session.regenerateId();
-                }
-                // continue
-                ctx.next();
-              });
-              return;
+      final HttpServerRequest request = ctx.request();
+      final String authorization = request.headers().get(HttpHeaders.AUTHORIZATION);
 
+      // if the provider supports PASSWORD flow or AUTH_CODE with JWT we can try to validate the Authorization header
+      if (authorization != null) {
+
+        String[] parts = authorization.split(" ");
+        if (parts.length == 2) {
+          final String scheme = parts[0], credentials = parts[1];
+
+          if (flow == PASSWORD && BASIC.matcher(scheme).matches()) {
+            final JsonObject login = new JsonObject();
+            final String decoded = new String(Base64.getDecoder().decode(parts[1]));
+            final int colonIdx = decoded.indexOf(":");
+
+            if (colonIdx != -1) {
+              login
+                .put("username", decoded.substring(0, colonIdx))
+                .put("password", decoded.substring(colonIdx + 1));
+            } else {
+              login
+                .put("username", decoded);
             }
-          } else {
-            ctx.response().putHeader("WWW-Authenticate", "Bearer error=\"invalid_token\"");
-            ctx.fail(401);
+
+            ((OAuth2Auth) authProvider).getToken(login, getToken -> {
+              if (getToken.failed()) {
+                ctx.response().putHeader("WWW-Authenticate", "Basic error=\"invalid_token\" error_message=\"" + getToken.cause().getMessage() + "\"");
+                ctx.fail(401);
+                return;
+              }
+
+              ctx.setUser(getToken.result());
+              Session session = ctx.session();
+              if (session != null) {
+                // the user has upgraded from unauthenticated to authenticated
+                // session should be upgraded as recommended by owasp
+                session.regenerateId();
+              }
+              // continue
+              ctx.next();
+            });
+            return;
+          }
+
+          if (flow == AUTH_CODE && supportJWT && BEARER.matcher(scheme).matches()) {
+            ((OAuth2Auth) authProvider).decodeToken(credentials, decodeToken -> {
+              if (decodeToken.failed()) {
```

Comment:
We don't really specify the reason why a token failed to decode, it can be either because the encoding was wrong, or because the signature did not match the payload. So we're returning 401 to differentiate from a unparsable token at the end where we return 400.